File: internal/application/get-service-commit-info.go
package application

import (
	"cchalop1.com/deploy/internal/api/service"
)

type ServiceCommitInfo struct {
	Hash      string `json:"hash"`
	Message   string `json:"message"`
	Author    string `json:"author"`
	Date      string `json:"date"`
	GithubUrl string `json:"githubUrl"`
}

// GetServiceCommitInfo récupère les informations du dernier commit déployé pour un service GitHub
func GetServiceCommitInfo(deployService *service.DeployService, serviceId string) (*ServiceCommitInfo, error) {
	// Récupérer le service par son ID
	service, err := deployService.DatabaseAdapter.GetServiceById(serviceId)
	if err != nil {
		return nil, err
	}

	// Vérifier que c'est un service GitHub
	if service.Type != "github_repo" {
		return nil, nil // Retourner nil pour les services non-GitHub sans erreur
	}

	// Vérifier si nous avons des informations de commit stockées
	if service.LastCommit.Hash == "" {
		return nil, nil // Pas encore de commit déployé
	}

	// Générer l'URL GitHub pour le commit
	githubUrl := deployService.GitAdapter.GetCommitGithubUrl(service.FullName, service.LastCommit.Hash)

	// Afficher seulement les 8 premiers caractères du hash
	shortHash := service.LastCommit.Hash
	if len(shortHash) > 8 {
		shortHash = shortHash[:8]
	}

	return &ServiceCommitInfo{
		Hash:      shortHash,
		Message:   service.LastCommit.Message,
		Author:    service.LastCommit.Author,
		Date:      service.LastCommit.Date,
		GithubUrl: githubUrl,
	}, nil
}

File: web/src/services/getServiceCommitInfo.ts
import { callApi } from "./api";

export type ServiceCommitInfo = {
  hash: string;
  message: string;
  author: string;
  date: string;
  githubUrl: string;
};

export async function getServiceCommitInfo(
  serviceId: string
): Promise<ServiceCommitInfo | null> {
  try {
    return await callApi<ServiceCommitInfo>(
      `/service/${serviceId}/commit-info`,
      "GET"
    );
  } catch (error) {
    // Si l'API retourne 204 (No Content), cela signifie que ce n'est pas un service GitHub
    if (error instanceof Error && error.message.includes("204")) {
      return null;
    }
    throw error;
  }
}
